### ISSUE TITLE
Missing metrics from PodMonitoring resource

### ISSUE THREAD

**Issue by elebioda** [GoogleCloudPlatform/prometheus-engine#152]
I setup kube-state-metrics and a PodMonitoring resource on the deployment

```
apiVersion: monitoring.googleapis.com/v1alpha1
kind: PodMonitoring
metadata:
  name: kube-state-metrics
  namespace: kube-state-metrics
spec:
  selector:
    matchLabels:
      app: kube-state-metrics
  endpoints:
  - port: http
    interval: 30s # Leave at 30 so we dont get charged 1 Billion dollaridoos 
```

But there are numerous metrics that are not being scraped and pushed to GMP

I made sure kube-state-metrics is publishing them.. for example our GMP is showing no metrics for a specific metric

```
PromQL Query: kube_deployment_status_replicas_available{deployment="tracking-order-api-order-tracking-api"}
Query duration: 0.6 sec
Series returned: 0
```

But when I check the metrics on kube-state-metrics they are showing as present on the /metrics endpoint

```
kube_deployment_status_replicas_available{namespace="prod",deployment="tracking-order-api-order-tracking-api"} 5
```

**Comment by lyanco** [comment]
kube-state-metrics monitors cluster state and is better handled by a ClusterPodMonitoring resource instead. Could you try the following and see if it works:

```
apiVersion: monitoring.googleapis.com/v1alpha1
kind: ClusterPodMonitoring
metadata:
  name: kube-state-metrics
spec:
  selector:
    matchLabels:
      app: kube-state-metrics
  endpoints:
  - port: http
    interval: 30s
  targetLabels:
    metadata: []
```

We'll soon have best-practice copy-and-paste CRs for kube state and other common exporters (e.g. node-exporter).

**Comment by elebioda** [comment]
> kube-state-metrics monitors cluster state and is better handled by a ClusterPodMonitoring resource instead. Could you try the following and see if it works as expected:
> 
> ```
> apiVersion: monitoring.googleapis.com/v1alpha1
> kind: ClusterPodMonitoring
> metadata:
>   name: kube-state-metrics
> spec:
>   selector:
>     matchLabels:
>       app: kube-state-metrics
>   endpoints:
>   - port: http
>     interval: 30s
>   targetLabels:
>     metadata: []
> ```
> 
> We'll soon have best-practice copy-and-paste CRs for kube state and other common exporters (e.g. node-exporter).

Im not sure I understand the benefits of the ClusterPodMonitoring vs the PodMonitoring. kube-state-metrics typically runs as a single replica as it would result in duplication of metrics. Could you elaborate on why ClusterPodMonitoring is better?

**Comment by elebioda** [comment]
> What I think is happening in your original situation is that the namespace in which kube-state is running (and which the PodMonitoring is installed in) is overriding the namespace label set on each individual kube-state metric. So queries that include the namespace "prod" are returning a null set.
> 
> I believe PodMonitorings hard-enforce the namespace in which they are installed, so they won't work at all for kube-state. ClusterPodMonitorings also try to add namespace, but by adding `targetLabels: metadata:[]`, you tell the system to not infer/overwrite the namespace label. So the namespace on each kube-state metric will be respected and your queries will work as expected.

Hmm ok I think I sort of get what you mean. I should have added this info, as I show in the example query, I am trying to query a specific deployment, this One deployment's metrics are missing for about 30 minutes rite now. Other deployment metrics are present on time of query 30 seconds, but not the one I showed above. Not sure if this changes your theory.

**Comment by lyanco** [comment]
I'd try using the ClusterPodMonitoring I added above and see if that fixes it. If it doesn't, then we can investigate further.

**Comment by elebioda** [comment]
The manifest you provided is not supported by the controller. Looks like there is a discrepancy between what the managed prometheus CRD it installs. I checked the CRD in our cluster and it does not contain the metadata field. 

```
              targetLabels:
                description: Labels to add to the Prometheus target for discovered
                  endpoints
                properties:
                  fromPod:
                    description: 'Labels to transfer from the Kubernetes Pod to Prometheus
                      target labels. In the case of a label mapping conflict: - Mappings
                      at the end of the array take precedence.'
                    items:
                      description: LabelMapping specifies how to transfer a label
                        from a Kubernetes resource onto a Prometheus target.
                      properties:
                        from:
                          description: Kubenetes resource label to remap.
                          type: string
                        to:
                          description: Remapped Prometheus target label. Defaults
                            to the same name as `From`.
                          type: string
                      required:
                      - from
                      type: object
                    type: array
                type: object
```

**Comment by lyanco** [comment]
Have you updated to the latest version of managed collection? v0.3.0

**Comment by lyanco** [comment]
You most likely installed the operator via `kubectl` in which case you are responsible for upgrading it. If you switch to using `gcloud` to install it then we will upgrade it for you automatically whenever there is a new version. Running `gcloud` in all your clusters should seamlessly override the `kubectl` installation and retain all your scrape configs.

See https://cloud.google.com/stackdriver/docs/managed-prometheus/setup-managed#config-mgd-collection

We just published this a week or so ago.

**Comment by elebioda** [comment]
I enabled it through the UI. I applied nothing through kubectl.

**Comment by elebioda** [comment]
<img width="709" alt="Screen Shot 2022-03-04 at 10 38 18 AM" src="https://user-images.githubusercontent.com/92803394/156803189-590bc0ad-4ac8-40f0-af03-a7afea96bcc4.png">

**Comment by lyanco** [comment]
Looks like we haven't bumped the version internally yet. We're doing that now, should roll out over the course of the next week.

Thanks for calling this out.

**Comment by elebioda** [comment]
Gotcha thanks for the update. So is It not possible to test your theory without the metadata field? Or will just the clusterpodmonitor suffice?

**Comment by lyanco** [comment]
In the meantime I believe if you comment out `targetLabels: metadata: []` then ClusterPodMonitoring should work; inferring namespace for ClusterPodMonitoring was only added in 0.3.0 so there shouldn't be anything you need to disable.

**Comment by elebioda** [comment]
The metrics are still delayed by about 30 minutes for very specific deployments for the metric: kube_deployment_status_replicas_available

**Comment by elebioda** [comment]
Im able to resolve this delay by just restarting the collector daemonset, but this is obviously not the solution.

**Comment by elebioda** [comment]
Hmmm, interesting, I am not sure if this solved the problem or the other team on GCP support fixed something, but the metrics have caught up. Will post an update Monday if they delay happens again. It seems to fix itself and then the delay shows up again a few days later.

**Comment by elebioda** [comment]
Update: Metrics are now consistent and always up to date with the use of ClusterPodMonitoring thank you!

**Comment by lyanco** [comment]
Glad it's working! Please be sure to re-apply the CR with `targetLabels: metadata: []` in the future (once your collection version updates). Your current CR should be good until the next time you apply it.

**Comment by lyanco** [comment]
One more question for you - where is your managed Prometheus collector running? Is this running within GCP, on-prem, or in some other Cloud?

**Comment by lyanco** [comment]
Update: We investigated some more and traced this down to a bug in our queueing/buffering logic. After any blip which caused a queue to form due to a missed sending cycle, the queue would only send one point per cycle, meaning it would never be able to catch up with the present time. Your fix was due to your restarting the collector and flushing the queues, but the root cause remains in this current release.

We're cutting a new release with the fix ASAP. It should gradually roll out to GKE over the next week or two.
